### ISSUE TITLE
POODLE TLS

### ISSUE THREAD

**Issue by drwetter** [drwetter/testssl.sh#43]
implement

https://gist.github.com/singe/f433c54f134a9390214e
https://www.imperialviolet.org/binary/scanpadding.go
https://www.imperialviolet.org/binary/poodle-tls-go.patch

**Comment by AntoineGalland** [comment]
Hi,
Any idea of the milestone that will support this feature?
Ivan Ristić has added a test for this issue to his excellent scanner at SSLLabs. Affected sites will have their grade set to F and will report “This server is vulnerable to the POODLE attack against TLS servers”. https://www.imperialviolet.org/2014/12/08/poodleagain.html
Is it possible at least that your script say that the POODLE TLS is not yet tested in order to make a distinction with PODDLE SSL which is tested?
It is confusing when we compare to SSLabs, especially with the grade F on the PODDLE TLS
https://community.qualys.com/blogs/securitylabs/2014/12/08/poodle-bites-tls
Regards

**Comment by drwetter** [comment]
Hi Antoine,

Am 04/30/2015 um 09:41 PM schrieb AntoineGalland:
> Hi,
> Any idea of the milestone that will support this feature?

I intend to plan soon a stable release 2.4. Soon= next 2-3 weeks.

Don't know it'll be possible to implement that.

> Ivan Ristić has added a test for this issue to his excellent scanner at SSLLabs. Affected sites
> will have their grade set to F and will report “This server is vulnerable to the POODLE attack
> against TLS servers”. https://www.imperialviolet.org/2014/12/08/poodleagain.html

I lack a) test cases for this and b) a solid idea how to test it.

a) is not easy as I don't have broken Cisco/Juniper/F5 hardware here to test against
   and SSLlabs' code is proprietary. If anybody knows a broken piece of HW publicly
   available send me a pm.

b) I thought I had a PoC last year but ssllabs had sometimes different opinions
   on the results.

Everybody is invited to implement POODLE/TLS in testssl.sh -- last but not least that's
the idea of Open Source Software.

> Is it possible at least that your script says that the POODLE TLS is not yet tested in order to
> make a distinction with PODDLE SSL which is tested?
>
> It is confusing when we compare to SSLabs, especially with the grade F on the PODDLE TLS
> https://community.qualys.com/blogs/securitylabs/2014/12/08/poodle-bites-tls

Currently the test says "POODLE, SSL (CVE-2014-3566)". How can this be clearer? I am
not tempted to add a line "POODLE, TLS:  not yet implemented" as this is obvious, isn't it?


Cheers, Dirk

**Comment by PeterMosmans** [comment]
Notes on implementation:
I'm quoting Ivan Ristic here (see https://community.qualys.com/thread/14494) on how Qualys detects CVE-2014-8730:
*"It's detected by attempting to negotiate a connection with the server, but using broken CBC padding. If the handshake succeeds, the server is considered vulnerable. We also submit a HTTP request and expect a response."*

If nobody else has (yet) jumped on this issue I could take a look at it somewhere in September or so ?

**Comment by drwetter** [comment]
That would be cool!

As far as I am concerned I haven't gotten a complete understanding how the padding should be implemented but I assume that "just using openssl" is not good enough.

The go code seems to be one good starting point but I am missing the whole code (diff is not enough). Another alternative would be https://github.com/exploresecurity/test_poodle_tls/blob/master/test_poodle_tls.py, but somebody needs to dig into the tlslite code probably.

**Comment by PeterMosmans** [comment]
Ok @drwetter , I'm moving this issue to my list for September, please label accordingly.
Note: no guarantees of course :)

Cheers,

Peter

**Comment by drwetter** [comment]
Very cool, @PeterMosmans .

Would love to put you in as an assignee but it seems that either github or my browser is not allowing to do so. Can't you do that?

Cheers, Dirk

**Comment by PeterMosmans** [comment]
Nope, the assignee button is greyed out, and I can't change the labels.
I'll guess we both have to "old-fashioned" remember it now :smile:

**Comment by drwetter** [comment]
yup :-)

**Comment by drwetter** [comment]
Hi Peter,

ping && just checking ;-)

Cheers, Dirk

**Comment by PeterMosmans** [comment]
Thanks for the ping @drwetter . It's still on the TODO list, but the September target hasn't been, err, reached. I'm still game, but work kinda sits in the way of working on this currently.
Bottom line: if anyone wants to do it, please go for it! If not, I'll pick this up in the foreseeable future.

**Comment by drwetter** [comment]
Thx for the prompt reply, @PeterMosmans !

The before-christmas-time is quite busy here, too.

Cheer, Dirk

**Comment by PeterMosmans** [comment]
OK, I looked at the available solutions. For a proper test, one needs not only to simulate the handshake, but actually establish a TLS connection and sent a HTTP request over it. This cannot be done in Bash, which means either using an external tool (e.g. using tlslite / Python), or a patched version of openssl. As we already use a custom fork of openssl this sounds like the easiest way to go. I made some changes to test it locally, but actually implementing it with proper s_client flags and the likes will take time, as I do *not* want to break the original functionality. I'd like to have the openssl 1.0.2-chacha fork as vanilla as can be. So, @drwetter , I'm going to implement the broken functionality in my fork, and this will mean that it will take even more time :wink:

**Comment by drwetter** [comment]
Am 01/19/2016 um 03:12 AM schrieb Peter Mosmans:
> OK, I looked at the available solutions. For a proper test, one needs not only to simulate the
> handshake, but actually establish a TLS connection and sent a HTTP request over it. This cannot
> be done in Bash, 

NACK :-) see below.

> which means either using an external tool (e.g. using tlslite / Python), or a
> patched version of openssl.

To me there are two solutions, too. The latter one I was thinking on, too. For the first
though: To be honest I do not like to introduce a foreign language library. This project
does everything for portability reasons/ease of installation in bash + openssl (with a tendency
doing more in bash sockets in). Momentarily it should not change.

Prove me wrong but there is a chance that it'll can be achieved by bash only. I forgot the details
of POODLE-TLS but we have

* a bash-based socket interface working until the ServerHello (including parser of it)
* openssl which establishes a full handshake where we can send anything over the
  encrypted channel what we like

The latter is being used for determining all HTTP headers and Secure Client-Initiated
Renegotiation. In the latest versions I switched from an EOF-based mechanism to a pipe-based,
so maybe this is not that obvious anymore. Hint: in the recent master You need just to grep for
"| $OPENSSL s_client". For previous versions I did it different:
https://github.com/drwetter/testssl.sh/blob/2.4/testssl.sh#L449-L471

Over those connections (EOF/pipe) we can send anything we want -- also raw bytes if we want to.

> As we already use a custom fork of openssl this sounds like the
> easiest way to go. I made some changes to test it locally, but actually implementing it with
> proper s_client flags and the likes will take time, as I do /not/ want to break the original
> functionality. I'd like to have the openssl 1.0.2-chacha fork as vanilla as can be. So,
> @drwetter <https://github.com/drwetter> , I'm going to implement the broken functionality in my
> fork, and this will mean that it will take even more time :wink:


Maybe my approach is faster? I'll be happy to give you a hand if you give me the details what
is best to send over the channel.


Cheers, Dirk

**Comment by PeterMosmans** [comment]
@drwetter , openssl sets the padding of TLS records 'hardcoded' in the source to the correct value, before sending the packets to the server.
That's what I patched for SSLv3 and TLSv1, to create a POODLE check using the openssl binary, where I b0rk the padding "manually" and to check the response of the server. So, you're messing with the transport layer, not stuff that's being sent _over_ the transport layer. For a proper POODLE test, you need to modify the padding _after_ a succesful handshake, after the encrypted channel has been set up. Is that possible using Bash ?

**Comment by drwetter** [comment]
That is what I thought initially but probably I misunderstood you saying "but actually establish a TLS connection and sent a HTTP request over it".

To answer your question: Everything is possible with bash, it's just a matter of effort. Details to follow.
